src/rust/engine/workunit_store/src/lib.rs

```diff
@@ -44,22 +44,37 @@ pub struct WorkUnit {
 
 #[derive(Clone, Default)]
 pub struct WorkUnitStore {
-  workunits: Arc<Mutex<HashSet<WorkUnit>>>,
+  workunits: Arc<Mutex<Vec<WorkUnit>>>,
```

Comment:
A part of me thinks that this Vec and this index belong in the same struct, bundled inside the same `Arc<Mutex>`. I can imagine how `get_workunits` would be complicated to implement, though; so if you're pretty convinced there is no easy way to do this, I'm OK with this as the code smell is pretty self-contained inside this struct anyways.

Reply:
Just having the one `Arc<Mutex<_>>` would make it impossible to deadlock, and would tie together the atomicity of lookup. I agree that if an `InnerWorkUnitStore` is hard to implement, I wouldn't bother, but I suspect it's not too bad...